tree/ntuple/v7/src/RPageStorageS3.cxx

```diff
@@ -0,0 +1,422 @@
+   WriteNTupleFooter(zipBuffer.get(), szZipFooter, szFooter);
+   WriteNTupleAnchor();
+}
+
+void ROOT::Experimental::Detail::RPageSinkS3::WriteNTupleHeader(
+		const void *data, size_t nbytes, size_t lenHeader)
+{
+   fS3Handle->WriteObject(kOidHeader, data, nbytes);
+   fNTupleAnchor.fLenHeader = lenHeader;
+   fNTupleAnchor.fNBytesHeader = nbytes;
+}
+
+void ROOT::Experimental::Detail::RPageSinkS3::WriteNTupleFooter(
+		const void *data, size_t nbytes, size_t lenFooter)
+{
+   fS3Handle->WriteObject(kOidFooter, data, nbytes);
+   fNTupleAnchor.fLenFooter = lenFooter;
+   fNTupleAnchor.fNBytesFooter = nbytes;
+}
+
+void ROOT::Experimental::Detail::RPageSinkS3::WriteNTupleAnchor() {
+   const auto ntplSize = RS3NTupleAnchor::GetSize();
+   auto buffer = std::make_unique<unsigned char[]>(ntplSize);
+   fNTupleAnchor.Serialize(buffer.get());
+   fS3Handle->WriteObject(kOidAnchor, buffer.get(), ntplSize);
+}
+
+ROOT::Experimental::Detail::RPage
+ROOT::Experimental::Detail::RPageSinkS3::ReservePage(ColumnHandle_t columnHandle, std::size_t nElements)
+{
+   if (nElements == 0)
+      nElements = fOptions->GetNElementsPerPage();
+   auto elementSize = columnHandle.fColumn->GetElement()->GetSize();
+   return fPageAllocator->NewPage(columnHandle.fId, elementSize, nElements);
+}
+
+void ROOT::Experimental::Detail::RPageSinkS3::ReleasePage(RPage &page)
+{
+   fPageAllocator->DeletePage(page);
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPage ROOT::Experimental::Detail::RPageAllocatorS3::NewPage(
+   ColumnId_t columnId, void *mem, std::size_t elementSize, std::size_t nElements)
+{
+   RPage newPage(columnId, mem, elementSize * nElements, elementSize);
+   newPage.TryGrow(nElements);
+   return newPage;
+}
+
+void ROOT::Experimental::Detail::RPageAllocatorS3::DeletePage(const RPage& page)
+{
+   if (page.IsNull())
+      return;
+   delete[] reinterpret_cast<unsigned char *>(page.GetBuffer());
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+
+
+ROOT::Experimental::Detail::RPageSourceS3::RPageSourceS3(
+   std::string_view ntupleName, std::string_view uri, const RNTupleReadOptions &options)
+   : RPageSource(ntupleName, options)
+   , fPageAllocator(std::make_unique<RPageAllocatorS3>())
+   , fPagePool(std::make_shared<RPagePool>())
+   , fUri(uri)
+{
+   RPageSource::fDecompressor = std::make_unique<RNTupleDecompressor>();
+   EnableDefaultMetrics("RPageSourceS3");
+   fS3Handle = std::make_unique<RS3Handle>(ntupleName, fUri);
+}
+
+ROOT::Experimental::Detail::RPageSourceS3::~RPageSourceS3() = default;
+
+ROOT::Experimental::RNTupleDescriptor ROOT::Experimental::Detail::RPageSourceS3::AttachImpl()
+{
+   R__ASSERT(fS3Handle);
+   RNTupleDescriptorBuilder descBuilder;
+
+   RS3NTupleAnchor ntpl;
+   std::vector<char> buf;
+   buf.reserve(RS3NTupleAnchor::GetSize());
+   fS3Handle->ReadObject(kOidAnchor, buf);
```

Comment:
If we happen to use `DavFile::readPartial()`, we should be able to provide a plain `char[]` buffer here.

(The same applies for the rest below).